--- src/com/nutiteq/location/NutiteqLocationMarker.java	2009-06-03 17:34:46.000000000 +0200
+++ src/com/nutiteq/location/NutiteqLocationMarker.java	2011-01-24 18:09:20.000000000 +0100
@@ -15,16 +15,16 @@
  * indicating connection state (either connected or connection lost).
  */
 public class NutiteqLocationMarker extends TimerTask implements LocationMarker {
+  protected BasicMapComponent mapComponent;
+  protected LocationSource locationSource;
+  protected MapPos mapPosition;
+  protected final Placemark placemarkConnected;
+  protected boolean track;
+  protected WgsPoint lastWgsLocation;
-  private WgsPoint lastWgsLocation;
-  private BasicMapComponent mapComponent;
-  private MapPos mapPosition;
-  private final Placemark placemarkConnected;
   private final int updateInterval;
   private boolean running;
-  private boolean track;
   private final Placemark placemarkConnectionLost;
   private final int lastStatus;
-  private LocationSource locationSource;
   //TODO jaanus : check this timer stuff
   private final Timer timer = new Timer();
 
@@ -103,7 +103,7 @@
     }
   }
 
-  private void update() {
+  protected void update() {
     if (mapComponent == null || lastWgsLocation == null) {
       return;
     }
--- src/javax/microedition/lcdui/Image.java	2010-06-11 20:16:18.000000000 +0200
+++ src/javax/microedition/lcdui/Image.java	2011-01-18 14:39:40.000000000 +0100
@@ -10,7 +10,7 @@
 public class Image {
   private final Bitmap bitmap;
 
-  private Image(final Bitmap bitmap) {
+  public Image(final Bitmap bitmap) {
     this.bitmap = bitmap;
   }
 
--- src/com/nutiteq/BasicMapComponent.java	2011-01-25 14:42:16.000000000 +0100
+++ src/com/nutiteq/BasicMapComponent.java	2011-02-01 10:26:05.000000000 +0100
@@ -187,9 +187,9 @@
   private boolean isMapComplete;
   private long lastZoomCall;
 
-  private final Vector changedAreas = new Vector();
+  private final Vector<Rectangle> changedAreas = new Vector<Rectangle>();
 
-  private LocationSource locationSource;
+  protected LocationSource locationSource;
   private GeoMap[] tileSearchStrategy;
 
   private final LicenseKeyCheck licenseKeyCheck;
