--- src/com/nutiteq/cache/CacheIndexDatabaseHelper.java	2009-06-07 17:39:20.000000000 +0200
+++ src/com/nutiteq/cache/CacheIndexDatabaseHelper.java	2010-11-22 15:30:56.000000000 +0100
@@ -20,7 +20,7 @@
   private final Context ctx;
   private DatabaseHelper databaseHelper;
   private final String databaseName;
-  private SQLiteDatabase database;
+  public SQLiteDatabase database;
 
   private static final String KEY_ID = "id";
   private static final String KEY_CACHE_KEY = "cache_key";
