--- src/com/nutiteq/location/NutiteqLocationMarker.java	2009-06-03 17:34:46.000000000 +0200
+++ src/com/nutiteq/location/NutiteqLocationMarker.java	2011-01-24 18:09:20.000000000 +0100
@@ -15,16 +15,16 @@
  * indicating connection state (either connected or connection lost).
  */
 public class NutiteqLocationMarker extends TimerTask implements LocationMarker {
+  protected BasicMapComponent mapComponent;
+  protected LocationSource locationSource;
+  protected MapPos mapPosition;
+  protected final Placemark placemarkConnected;
+  protected boolean track;
   private WgsPoint lastWgsLocation;
-  private BasicMapComponent mapComponent;
-  private MapPos mapPosition;
-  private final Placemark placemarkConnected;
   private final int updateInterval;
   private boolean running;
-  private boolean track;
   private final Placemark placemarkConnectionLost;
   private final int lastStatus;
-  private LocationSource locationSource;
   //TODO jaanus : check this timer stuff
   private final Timer timer = new Timer();
 
@@ -103,7 +103,7 @@
     }
   }
 
-  private void update() {
+  protected void update() {
     if (mapComponent == null || lastWgsLocation == null) {
       return;
     }
