--- src/com/nutiteq/BasicMapComponent.java
+++ src/com/nutiteq/BasicMapComponent.java
@@ -137,7 +137,7 @@ public class BasicMapComponent extends BaseMapComponent implements MapTilesReque
  protected GeoMap displayedMap;

  private KmlServicesHandler kmlServicesHandler;
-  private com.nutiteq.cache.Cache networkCache;
+  protected com.nutiteq.cache.Cache networkCache;

  private final TasksRunner taskRunner;

--- src/com/nutiteq/BasicMapComponent.java
+++ src/com/nutiteq/BasicMapComponent.java
@@ -1964,10 +1964,19 @@ public class BasicMapComponent extends BaseMapComponent implements MapTilesReque
       return;
     }
 
-    mappingStarted = false;
-
-    panning.quit();
-
+    if (screenBuffer != null) {
+        screenBuffer.getBitmap().recycle();
+        screenBuffer = null;
+    }
+    screenBufferGraphics = null;
+    screenCache.deinitialize();
+
+    mappingStarted = false;
+
+    if (panning != null) {
+        panning.quit();
+    }
+
     if (locationSource != null) {
       locationSource.quit();
     }
--- src/com/nutiteq/components/ImageBuffer.java
+++ src/com/nutiteq/components/ImageBuffer.java
@@ -48,6 +48,11 @@ public class ImageBuffer {
   }
 
   public void clean() {
+    for (int i = 0; i < bufferImages.length; i++) {
+      if (bufferImages[i] != null) {
+        bufferImages[i].getBitmap().recycle();
+      }
+    }
     bufferImages = null;
     bufferGraphics = null;
   }
--- src/com/nutiteq/components/MapTile.java
+++ src/com/nutiteq/components/MapTile.java
@@ -123,7 +123,7 @@ public class MapTile {
    * Not part of public API
    */
   public void setImagesData(final byte[][] data) {
-    if (data == null || data.length == 0 || data[0].length == 0) {
+    if (data == null || data.length == 0 || data[0] == null || data[0].length == 0) {
       failCount++;
       Log.error("Could not retrieve " + getIDString());
       requestor.tileRetrieved(this);
@@ -169,12 +169,16 @@ public class MapTile {
           }
         }
         imageGraphics.drawImage(tmp, 0, 0, Graphics.TOP | Graphics.LEFT);
+        tmp.getBitmap().recycle();
+        tmp = null;
       }
 
       if (overlayData != null) {
         for (int i = 0; i < overlayData.length; i++) {
-          final Image tmp = Image.createImage(overlayData[i], 0, overlayData[i].length);
+          Image tmp = Image.createImage(overlayData[i], 0, overlayData[i].length);
           imageGraphics.drawImage(tmp, 0, 0, Graphics.TOP | Graphics.LEFT);
+          tmp.getBitmap().recycle();
+          tmp = null;
         }
       }
       return tileImage;
--- src/com/nutiteq/utils/Utils.java
+++ src/com/nutiteq/utils/Utils.java
@@ -184,13 +184,15 @@ public class Utils {
     throw new IllegalArgumentException("Do not know the log2 from " + x);
   }
 
-  public static Image resizeImageAndCopyPrevious(final int newWidth, final int newHeight, final Image resized) {
+  public static Image resizeImageAndCopyPrevious(final int newWidth, final int newHeight, Image resized) {
     // TODO jaanus : if new is smaller can optimize with
     // createImage(Image image, int x, int y, int width, int height, int
     // transform)
     final Image result = Image.createImage(newWidth, newHeight);
     final Graphics g = result.getGraphics();
     g.drawImage(resized, (newWidth - resized.getWidth()) / 2, (newHeight - resized.getHeight()) / 2, Graphics.TOP | Graphics.LEFT);
+    resized.getBitmap().recycle();
+    resized = null;
     return result;
   }
 
