--- maps_lib/android/src/com/nutiteq/android/RepaintHandler.java	2009-10-28 16:28:36.000000000 +0100
+++ maps_lib-patched/src/com/nutiteq/android/RepaintHandler.java	2010-11-16 16:27:48.000000000 +0100
@@ -12,7 +12,9 @@
 
   @Override
   public void handleMessage(final Message msg) {
-    mapView.invalidate();
+    if (mapView != null) {
+        mapView.invalidate();
+    }
     super.handleMessage(msg);
   }
 
