--- src/com/mgmaps/cache/ScreenCache.java
+++ src/com/mgmaps/cache/ScreenCache.java
@@ -105,6 +105,7 @@ public class ScreenCache {
     if (images[i] == null) {
       //TODO jaanus : why does this sometimes happen with streamed tiles?
       Log.debug(">>>>>>>>>>>>>>>>>>>>>>> SC: null image!");
+      valid[i] = false; //FIXME
       images[i] = tiles[i].getMap().getMissingTileImage();
     }
     final int left = tiles[i].getX() - centerCopy.getX() + screenCenterX;
@@ -121,7 +122,7 @@ public class ScreenCache {
    */
   public int find(final MapTile t) {
     for (int i = 0; i < size; i++) {
-      if (valid[i] && tiles[i].equals(t)) {
+      if (valid[i] && tiles[i] != null && tiles[i].equals(t)) {
         return i;
       }
     }
@@ -181,10 +182,15 @@ public class ScreenCache {
       if (!valid[i]) {
         continue;
       }
+      // if found, do not remove it
       if (tiles[i].equals(t)) {
         found = i;
+        break;
       } else if (!tiles[i].isVisible(mp, displayedMap, screenCenterX, screenCenterY)) {
-        // if found, do not remove it
+        // Clean stuff in memory
+        if (images[i] != null) {
+            images[i].getBitmap().recycle();
+        }
         valid[i] = false;
         tiles[i] = null;
         images[i] = null;
@@ -206,6 +212,9 @@ public class ScreenCache {
   public void setImageProcessor(final ImageProcessor processor) {
     imageProcessor = processor;
     for (int i = 0; i < size; i++) {
+      if (images[i] != null) {
+        images[i].getBitmap().recycle();
+      }
       valid[i] = false;
       tiles[i] = null;
       images[i] = null;
