--- src/com/mgmaps/cache/ScreenCache.java	2010-11-23 10:44:36.000000000 +0100
+++ src/com/mgmaps/cache/ScreenCache.java	2011-01-31 15:45:34.564297534 +0100
@@ -59,6 +59,17 @@
     resize(n);
   }
 
+  public void deinitialize() {
+    for (int i = 0; i < images.length; i++) {
+      if (images[i] != null) {
+        images[i].getBitmap().recycle();
+      }
+    }
+    tiles = null;
+    images = null;
+    valid = null;
+  }
+
   /**
    * Resize the screen cache when switching full screen.
    * 
@@ -100,9 +111,10 @@
    *          copy of the map center, used for synchronization
    */
   public void paint(final Graphics g, final int i, final MapPos centerCopy, final int screenCenterX, final int screenCenterY) {
-    if (images[i] == null) {
+    if (images[i] == null || !valid[i]) {
       //TODO jaanus : why does this sometimes happen with streamed tiles?
       Log.debug(">>>>>>>>>>>>>>>>>>>>>>> SC: null image!");
+      valid[i] = false; //FIXME
       images[i] = tiles[i].getMap().getMissingTileImage();
     }
     final int left = tiles[i].getX() - centerCopy.getX() + screenCenterX;
@@ -119,7 +131,7 @@
    */
   public int find(final MapTile t) {
     for (int i = 0; i < size; i++) {
-      if (valid[i] && tiles[i].equals(t)) {
+      if (valid[i] && images[i] != null && tiles[i] != null && tiles[i].equals(t)) {
         return i;
       }
     }
@@ -204,6 +216,9 @@
   public void setImageProcessor(final ImageProcessor processor) {
     imageProcessor = processor;
     for (int i = 0; i < size; i++) {
+      if (images[i] != null) {
+        images[i].getBitmap().recycle();
+      }
       valid[i] = false;
       tiles[i] = null;
       images[i] = null;
